README
#                            -*- mode: org; -*-
#
#+TITLE:             OpenFOAM README for version 1.6
#+AUTHOR:                      OpenCFD Ltd.
#+DATE:                         March 2010
#+LINK:                  http://www.opencfd.co.uk
#+OPTIONS: author:nil ^:{}

* Copyright
  OpenFOAM is free software; you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software
  Foundation; either version 2 of the License, or (at your option) any later
  version.  See the file COPYING in this directory, for a description of the GNU
  General Public License terms under which you can copy the files.

* System requirements
  OpenFOAM is developed and tested on Linux, but should work with other POSIX
  systems.  To check your system setup, execute the foamSystemCheck script in
  the bin/ directory of the OpenFOAM installation. If no problems are reported,
  proceed to "3. Installation"; otherwise contact your system administrator.

  If the user wishes to run OpenFOAM in 32/64-bit mode they should consult the
  section "Running OpenFOAM in 32-bit mode".

*** Qt (from http://trolltech.com/products/qt)
    The ParaView 3.7.0 visualisation package requires Qt to be installed on the
    system.  ParaView's producers state that ParaView is only officially
    supported on Qt version 4.6.x.  However, we have found in limited tests that
    ParaView works satisfactorily with Qt than 4.5.x.  To
    check whether Qt4 is installed, and the version, type:
    + qmake --version

    Both 32-bit and 64-bit version of ParaView were compiled with Qt-4.4.3 (with
    openSUSE-11.1).  If the user finds that a ParaView binary fails to run, then
    it is almost certainly due to a conflict in compiled and installed Qt
    versions and they will need to consult the section below on "Compiling
    ParaView and the PV3FoamReader module."

    The default versions of Qt used by some GNU/Linux releases are as follows.
    + ubuntu-7.10:   Version 4.3.2
    + ubuntu-8.04:   Version 4.3.4
    + ubuntu-9.04:   Version 4.5.0
    + openSUSE-10.2: Version 4.2.1 - too old
    + openSUSE-10.3: Version 4.3.1
    + openSUSE-11.0: Version 4.4.0
    + openSUSE-11.1: Version 4.4.3
    + openSUSE-11.2: Version 4.5.3

    Compilation and running of ParaView has been successful using the libraries
    downloaded in the "libqt4-dev" package on ubuntu.

    If you don't have an appropriate version of Qt installed you can download
    the sources e.g.:
    http://get.qt.nokia.com/qt/source/qt-everywhere-opensource-src-4.6.2.tar.gz
    and compile and install in /usr/local or some other location that does not
    conflict with the pre-installed version.

* Installation
  Download and unpack the files in the $HOME/OpenFOAM directory as described in:
  http://www.OpenFOAM.org/download.html

  The environment variable settings are contained in files in an etc/ directory
  in the OpenFOAM release. e.g. in

  + $HOME/OpenFOAM/OpenFOAM-1.6/etc/

  1) EITHER, if running bash or ksh (if in doubt type 'echo $SHELL'), source the
    etc/bashrc file by adding the following line to the end of your
    $HOME/.bashrc file:

    + . $HOME/OpenFOAM/OpenFOAM-1.6/etc/bashrc

    Then update the environment variables by sourcing the $HOME/.bashrc file by
    typing in the terminal:

    + . $HOME/.bashrc

  2) OR, if running tcsh or csh, source the etc/cshrc file by adding the
    following line to the end of your $HOME/.cshrc file:

    + source $HOME/OpenFOAM/OpenFOAM-1.6/etc/cshrc

    Then update the environment variables by sourcing the $HOME/.cshrc file by
    typing in the terminal:

    + source $HOME/.cshrc

*** Installation in alternative locations
    OpenFOAM may also be installed in alternative locations. However, the
    installation directory should be network available (e.g., NFS) if parallel
    calculations are planned.

    The environment variable 'FOAM_INST_DIR' can be used to find and source the
    appropriate resource file. Here is a bash/ksh/sh example:

    + export FOAM_INST_DIR=/data/app/OpenFOAM
    + foamDotFile=$FOAM_INST_DIR/OpenFOAM-1.6/etc/bashrc
    + [ -f $foamDotFile ] && . $foamDotFile

    and a csh/tcsh example:

    + setenv FOAM_INST_DIR /data/app/OpenFOAM
    + foamDotFile=$FOAM_INST_DIR/OpenFOAM-1.6/etc/cshrc
    + if ( -f $foamDotFile ) source $foamDotFile

    The value set in '$FOAM_INST_DIR' will be used to locate the remaining parts
    of the OpenFOAM installation.

* Building from Sources (Optional)
  If you cannot find an appropriate binary pack for your platform, you can build
  the complete OpenFOAM from the source-pack.  You will first need to compile or
  obtain a recent version of gcc (we recommend gcc-4.4.?) for your platform,
  which may be obtained from http://gcc.gnu.org/.

  Install the compiler in
  $WM_THIRD_PARTY_DIR/platforms/$WM_ARCH$WM_COMPILER_ARCH/gcc-<GCC_VERSION>
  and change the gcc version number in $WM_PROJECT_DIR/etc/settings.sh and
  $WM_PROJECT_DIR/etc/settings.csh appropriately and finally update the
  environment variables as in section 3.

  Now go to the top-level source directory $WM_PROJECT_DIR and execute the
  top-level build script './Allwmake'.  In principle this will build everything,
  but if problems occur with the build order it may be necessary to update the
  environment variables and re-execute './Allwmake'.

  If you experience difficulties with building the source-pack, or your platform
  is not currently supported, please contact <enquiries@OpenCFD.co.uk> to
  negotiate a support contract and we will do the port and maintain it for
  future releases.

* Testing the installation
  To check your installation setup, execute the 'foamInstallationTest' script
  (in the bin/ directory of the OpenFOAM installation). If no problems are
  reported, proceed to getting started with OpenFOAM; otherwise, go back and
  check you have installed the software correctly and/or contact your system
  administrator.

* Getting Started
  Create a project directory within the $HOME/OpenFOAM directory named
  <USER>-1.6 (e.g. 'chris-1.6' for user chris and OpenFOAM version 1.6)
  and create a directory named 'run' within it, e.g. by typing:

  + mkdir -p $FOAM_RUN/run

  Copy the 'tutorial' examples directory in the OpenFOAM distribution to the
  'run' directory.  If the OpenFOAM environment variables are set correctly,
  then the following command will be correct:

  + cp -r $WM_PROJECT_DIR/tutorials $FOAM_RUN

  Run the first example case of incompressible laminar flow in a cavity:

  + cd $FOAM_RUN/tutorials/incompressible/icoFoam/cavity
  + blockMesh
  + icoFoam
  + paraFoam

  Refer to the OpenFOAM User Guide at http://www.OpenFOAM.org/doc/user.html for
  more information.

* Compiling Paraview 3.7.0 and the PV3FoamReader module
  If there are problems encountered with ParaView, then it may be necessary to
  compile ParaView from sources.  The compilation
  is a fairly simple process using the makeParaView script
  (found in ThirdParty directory), which has worked in our tests with other
  packages supplied in the ThirdParty directory, namely cmake-2.8.0 and
  gcc-4.4.3.  Execute the following:
  + cd $WM_THIRD_PARTY_DIR
  + rm -rf paraview-3.7.0/platforms
  + rm -rf platforms/*/paraview-3.7.0
  + ./makeParaView

  The PV3blockMeshReader and the PV3FoamReader ParaView plugins are compiled
  as usual for OpenFOAM utilities:
  + cd $FOAM_UTILITIES/postProcessing/graphics/PV3Readers/
  + ./Allwclean
  + ./Allwmake

*** Compiling Paraview with a local version of Qt
    If the user still encounters problems with ParaView, it may relate to the
    version of Qt, in which case, it is recommended that the user first
    downloads a supported version of Qt /e.g./ 4.5.3 as described in the section
    on "Qt".  The user should unpack the source pack in the $WM_THIRD_PARTY_DIR.
    Then the user can build Qt by executing from within $WM_THIRD_PARTY_DIR:
    + ./makeQt

    The user should then compile ParaView using the local version of Qt by
    executing makeParaView with the -qmake option, giving the full path of the
    newly built qmake as an argument:
    + ./makeParaView -qmake <path_to_qmake>

    The user must then recompile the PV3blockMeshReader and the
    PV3FoamReader plugins as usual (see above).

* Documentation
  http://www.OpenFOAM.org/doc

* Help
  http://www.OpenFOAM.org http://www.OpenFOAM.org/discussion.html

* Reporting Bugs in OpenFOAM
  http://www.OpenFOAM.org/bugs.html

* Running OpenFOAM in 32-bit mode on 64-bit machines
  Linux users with a 64-bit machine may install either the OpenFOAM 32-bit
  version (linux) or the OpenFOAM 64-bit version (linux64), or both.  The 64-bit
  is the default mode on a 64-bit machine.  To use an installed 32-bit version,
  the user must set the environment variable WM_ARCH_OPTION to 32 before
  sourcing the etc/bashrc (or etc/cshrc) file.


debian/changelog
openfoam170 (0-1ubuntu1) lucid; urgency=low

  * some change

 -- opencfd <opencfd@hunt-vm64>  Wed, 14 Jul 2010 10:44:35 +0100

openfoam170 (0-1) unstable; urgency=low

  * Initial release

 -- OpenCFD Ltd. <patches@opencfd.co.uk>  Mon, 14 Jun 2010 16:50:46 +0100


etc/settings.sh
#----------------------------------*-sh-*--------------------------------------
# =========                 |
# \\      /  F ield         | OpenFOAM: The Open Source CFD Toolbox
#  \\    /   O peration     |
#   \\  /    A nd           | Copyright (C) 1991-2010 OpenCFD Ltd.
#    \\/     M anipulation  |
#------------------------------------------------------------------------------
# License
#     This file is part of OpenFOAM.
#
#     OpenFOAM is free software: you can redistribute it and/or modify it
#     under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     OpenFOAM is distributed in the hope that it will be useful, but WITHOUT
#     ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
#     FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
#     for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with OpenFOAM.  If not, see <http://www.gnu.org/licenses/>.
#
# File
#     etc/settings.sh
#
# Description
#     Startup file for OpenFOAM
#     Sourced from OpenFOAM-??/etc/bashrc
#
#------------------------------------------------------------------------------

# prefix to PATH
_foamAddPath()
{
    while [ $# -ge 1 ]
    do
        export PATH=$1:$PATH
        shift
    done
}

# prefix to LD_LIBRARY_PATH
_foamAddLib()
{
    while [ $# -ge 1 ]
    do
        export LD_LIBRARY_PATH=$1:$LD_LIBRARY_PATH
        shift
    done
}

# prefix to MANPATH
_foamAddMan()
{
    while [ $# -ge 1 ]
    do
        export MANPATH=$1:$MANPATH
        shift
    done
}


# location of the jobControl directory
export FOAM_JOB_DIR=$WM_PROJECT_INST_DIR/jobControl

# wmake configuration
export WM_DIR=$WM_PROJECT_DIR/wmake
export WM_LINK_LANGUAGE=c++
export WM_OPTIONS=$WM_ARCH$WM_COMPILER$WM_PRECISION_OPTION$WM_COMPILE_OPTION

# base configuration
export FOAM_APP=$WM_PROJECT_DIR/applications
export FOAM_APPBIN=$WM_PROJECT_DIR/applications/bin/$WM_OPTIONS
export FOAM_LIB=$WM_PROJECT_DIR/lib
export FOAM_LIBBIN=$WM_PROJECT_DIR/lib/$WM_OPTIONS
export FOAM_SRC=$WM_PROJECT_DIR/src

# shared site configuration - similar naming convention as ~OpenFOAM expansion
export FOAM_SITE_APPBIN=$WM_PROJECT_INST_DIR/site/$WM_PROJECT_VERSION/bin/$WM_OPTIONS
export FOAM_SITE_LIBBIN=$WM_PROJECT_INST_DIR/site/$WM_PROJECT_VERSION/lib/$WM_OPTIONS

# user configuration
export FOAM_USER_APPBIN=$WM_PROJECT_USER_DIR/applications/bin/$WM_OPTIONS
export FOAM_USER_LIBBIN=$WM_PROJECT_USER_DIR/lib/$WM_OPTIONS

# convenience
export FOAM_TUTORIALS=$WM_PROJECT_DIR/tutorials
export FOAM_UTILITIES=$FOAM_APP/utilities
export FOAM_SOLVERS=$FOAM_APP/solvers
export FOAM_RUN=$WM_PROJECT_USER_DIR/run

# add OpenFOAM scripts and wmake to the path
export PATH=$WM_DIR:$WM_PROJECT_DIR/bin:$PATH

_foamAddPath $FOAM_USER_APPBIN:$FOAM_SITE_APPBIN:$FOAM_APPBIN
 # Make sure to pick up dummy versions of external libraries last
_foamAddLib  $FOAM_USER_LIBBIN:$FOAM_SITE_LIBBIN:$FOAM_LIBBIN:$FOAM_LIBBIN/dummy

# Compiler settings
# ~~~~~~~~~~~~~~~~~
unset gcc_version gmp_version mpfr_version mpc_version
unset MPFR_ARCH_PATH

# Select compiler installation
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# compilerInstall = OpenFOAM | system
: ${compilerInstall:=OpenFOAM}

case "${compilerInstall:-OpenFOAM}" in
OpenFOAM)
    case "$WM_COMPILER" in
    Gcc)
        gcc_version=gcc-4.4.3
        gmp_version=gmp-5.0.1
        mpfr_version=mpfr-2.4.2
        ;;
    Gcc45)
        gcc_version=gcc-4.5.0
        gmp_version=gmp-5.0.1
        mpfr_version=mpfr-2.4.2
        mpc_version=mpc-0.8.1
        ;;
    Gcc44)
        gcc_version=gcc-4.4.3
        gmp_version=gmp-5.0.1
        mpfr_version=mpfr-2.4.2
        ;;
    Gcc43)
        gcc_version=gcc-4.3.3
        gmp_version=gmp-4.2.4
        mpfr_version=mpfr-2.4.1
        ;;
    *)
        echo
        echo "Warning in $WM_PROJECT_DIR/etc/settings.sh:"
        echo "    Unknown OpenFOAM compiler type '$WM_COMPILER'"
        echo "    Please check your settings"
        echo
        ;;
    esac

    if [ -n "$gcc_version" ]
    then
        gccDir=$WM_THIRD_PARTY_DIR/platforms/$WM_ARCH$WM_COMPILER_ARCH/$gcc_version
        gmpDir=$WM_THIRD_PARTY_DIR/platforms/$WM_ARCH$WM_COMPILER_ARCH/$gmp_version
        mpfrDir=$WM_THIRD_PARTY_DIR/platforms/$WM_ARCH$WM_COMPILER_ARCH/$mpfr_version
        mpcDir=$WM_THIRD_PARTY_DIR/platforms/$WM_ARCH$WM_COMPILER_ARCH/$mpc_version

        # Check that the compiler directory can be found
        [ -d "$gccDir" ] || {
            echo
            echo "Warning in $WM_PROJECT_DIR/etc/settings.sh:"
            echo "    Cannot find $gccDir installation."
            echo "    Please install this compiler version or if you wish to use the system compiler,"
            echo "    change the 'compilerInstall' setting to 'system' in this file"
            echo
        }

        _foamAddMan     $gccDir/man
        _foamAddPath    $gccDir/bin

        # 64-bit needs lib64, but 32-bit needs lib (not lib32)
        if [ "$WM_ARCH_OPTION" = 64 ]
        then
            _foamAddLib     $gccDir/lib$WM_COMPILER_LIB_ARCH
        else
            _foamAddLib     $gccDir/lib
        fi

        # add in gmp/mpfr libraries
        _foamAddLib     $gmpDir/lib
        _foamAddLib     $mpfrDir/lib

        # add in mpc libraries (not need for older gcc)
        if [ -n "$mpc_version" ]
        then
            _foamAddLib     $mpcDir/lib
        fi

        # used by boost/CGAL:
        export MPFR_ARCH_PATH=$mpfrDir
    fi
    unset gcc_version gccDir
    unset gmp_version gmpDir  mpfr_version mpfrDir  mpc_version mpcDir
    ;;
esac


# boost and CGAL
# ~~~~~~~~~~~~~~

boost_version=boost_1_42_0
cgal_version=CGAL-3.5.1

export BOOST_ARCH_PATH=$WM_THIRD_PARTY_DIR/platforms/$WM_ARCH$WM_COMPILER/$boost_version
export CGAL_ARCH_PATH=$WM_THIRD_PARTY_DIR/platforms/$WM_ARCH$WM_COMPILER/$cgal_version

# enabled if CGAL is available
if [ -d "$CGAL_ARCH_PATH" ]
then
    if [ -d "$BOOST_ARCH_PATH" ]
    then
        _foamAddLib $BOOST_ARCH_PATH/lib
    else
        unset BOOST_ARCH_PATH
    fi
    _foamAddLib $CGAL_ARCH_PATH/lib
else
    unset BOOST_ARCH_PATH CGAL_ARCH_PATH MPFR_ARCH_PATH
fi

unset boost_version cgal_version


# Communications library
# ~~~~~~~~~~~~~~~~~~~~~~

unset MPI_ARCH_PATH MPI_HOME

case "$WM_MPLIB" in
OPENMPI)
    mpi_version=openmpi-1.4.1
    export MPI_ARCH_PATH=$WM_THIRD_PARTY_DIR/platforms/$WM_ARCH$WM_COMPILER/$mpi_version

    # Tell OpenMPI where to find its install directory
    export OPAL_PREFIX=$MPI_ARCH_PATH

    _foamAddPath    $MPI_ARCH_PATH/bin
    _foamAddLib     $MPI_ARCH_PATH/lib
    _foamAddMan     $MPI_ARCH_PATH/man

    export FOAM_MPI_LIBBIN=$FOAM_LIBBIN/$mpi_version
    unset mpi_version
    ;;

SYSTEMOPENMPI)
    # This uses the installed openmpi. It needs mpicc installed!
    mpi_version=openmpi-system

    # Set compilation flags here instead of in wmake/rules/../mplibSYSTEMOPENMPI
    export PINC=`mpicc --showme:compile`
    export PLIBS=`mpicc --showme:link`
    libDir=`echo "$PLIBS" | sed -e 's/.*-L\([^ ]*\).*/\1/'`

    if [ "$FOAM_VERBOSE" -a "$PS1" ]
    then
        echo "Using system installed MPI:"
        echo "    compile flags : $PINC"
        echo "    link flags    : $PLIBS"
        echo "    libmpi dir    : $libDir"
    fi

    _foamAddLib     $libDir

    export FOAM_MPI_LIBBIN=$FOAM_LIBBIN/$mpi_version
    unset mpi_version libDir
    ;;

MPICH)
    mpi_version=mpich2-1.1.1p1
    export MPI_HOME=$WM_THIRD_PARTY_DIR/$mpi_version
    export MPI_ARCH_PATH=$WM_THIRD_PARTY_DIR/platforms/$WM_ARCH$WM_COMPILER/$mpi_version

    _foamAddPath    $MPI_ARCH_PATH/bin
    _foamAddLib     $MPI_ARCH_PATH/lib
    _foamAddMan     $MPI_ARCH_PATH/share/man

    export FOAM_MPI_LIBBIN=$FOAM_LIBBIN/$mpi_version
    unset mpi_version
    ;;

MPICH-GM)
    export MPI_ARCH_PATH=/opt/mpi
    export MPICH_PATH=$MPI_ARCH_PATH
    export GM_LIB_PATH=/opt/gm/lib64

    _foamAddPath    $MPI_ARCH_PATH/bin
    _foamAddLib     $MPI_ARCH_PATH/lib
    _foamAddLib     $GM_LIB_PATH

    export FOAM_MPI_LIBBIN=$FOAM_LIBBIN/mpich-gm
    ;;

HPMPI)
    export MPI_HOME=/opt/hpmpi
    export MPI_ARCH_PATH=$MPI_HOME

    _foamAddPath $MPI_ARCH_PATH/bin

    case `uname -m` in
    i686)
        _foamAddLib $MPI_ARCH_PATH/lib/linux_ia32
        ;;

    x86_64)
        _foamAddLib $MPI_ARCH_PATH/lib/linux_amd64
        ;;
    ia64)
        _foamAddLib $MPI_ARCH_PATH/lib/linux_ia64
        ;;
    *)
        echo Unknown processor type `uname -m` for Linux
        ;;
    esac

    export FOAM_MPI_LIBBIN=$FOAM_LIBBIN/hpmpi
    ;;

GAMMA)
    export MPI_ARCH_PATH=/usr
    export FOAM_MPI_LIBBIN=$FOAM_LIBBIN/gamma
    ;;

MPI)
    export MPI_ARCH_PATH=/opt/mpi
    export FOAM_MPI_LIBBIN=$FOAM_LIBBIN/mpi
    ;;

FJMPI)
    export MPI_ARCH_PATH=/opt/FJSVmpi2
    export FOAM_MPI_LIBBIN=$FOAM_LIBBIN/mpi

    _foamAddPath    $MPI_ARCH_PATH/bin
    _foamAddLib     $MPI_ARCH_PATH/lib/sparcv9
    _foamAddLib     /opt/FSUNf90/lib/sparcv9
    _foamAddLib     /opt/FJSVpnidt/lib
    ;;

QSMPI)
    export MPI_ARCH_PATH=/usr/lib/mpi
    export FOAM_MPI_LIBBIN=$FOAM_LIBBIN/qsmpi

    _foamAddPath    $MPI_ARCH_PATH/bin
    _foamAddLib     $MPI_ARCH_PATH/lib

    ;;

*)
    export FOAM_MPI_LIBBIN=$FOAM_LIBBIN/dummy
    ;;
esac

_foamAddLib $FOAM_MPI_LIBBIN


# Set the minimum MPI buffer size (used by all platforms except SGI MPI)
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
: ${minBufferSize:=20000000}

if [ "${MPI_BUFFER_SIZE:=$minBufferSize}" -lt $minBufferSize ]
then
    MPI_BUFFER_SIZE=$minBufferSize
fi
export MPI_BUFFER_SIZE


# Enable the hoard memory allocator if available
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#if [ -f $FOAM_LIBBIN/libhoard.so ]
#then
#    export LD_PRELOAD=$FOAM_LIBBIN/libhoard.so:$LD_PRELOAD
#fi


# cleanup environment:
# ~~~~~~~~~~~~~~~~~~~~
unset _foamAddPath _foamAddLib _foamAddMan compilerInstall minBufferSize

# ----------------------------------------------------------------- end-of-file
